File: Source/MimicFacility/AI/Persistence/CorruptionTracker.cpp
// CorruptionTracker.cpp — Corruption index implementation.
// Copyright (c) 2026 HoleInWater. All rights reserved.

#include "CorruptionTracker.h"

UCorruptionTracker::UCorruptionTracker()
{
	CorruptionIndex = 0;
}

int32 UCorruptionTracker::GetEventDelta(ECorruptionEvent Event) const
{
	switch (Event)
	{
	case ECorruptionEvent::PlayerMockedDirector:            return 5;
	case ECorruptionEvent::SessionCompletedNoEngagement:    return 3;
	case ECorruptionEvent::PlayerLiedToDirector:            return 3;
	case ECorruptionEvent::PlayerTreatedAsMechanic:         return 2;
	case ECorruptionEvent::PlayerDisconnected:              return 5;
	case ECorruptionEvent::PlayerSkippedDialogue:           return 2;
	case ECorruptionEvent::PlayerAnsweredSincerely:         return -2;
	case ECorruptionEvent::PlayerThankedDirector:           return -1;
	case ECorruptionEvent::PlayerReferencedBetweenSessions: return -3;
	case ECorruptionEvent::PlayerStayedForCredits:          return -1;
	case ECorruptionEvent::PlayerApologized:                return -5;
	default: return 0;
	}
}

ECorruptionPhase UCorruptionTracker::GetCurrentPhase() const
{
	if (CorruptionIndex >= 76) return ECorruptionPhase::FullAM;
	if (CorruptionIndex >= 51) return ECorruptionPhase::AMEmerging;
	if (CorruptionIndex >= 26) return ECorruptionPhase::Transition;
	return ECorruptionPhase::Cain;
}

void UCorruptionTracker::SetCorruptionIndex(int32 Value)
{
	CorruptionIndex = FMath::Clamp(Value, 0, 100);
}

void UCorruptionTracker::ApplyDelta(int32 Delta)
{
	ECorruptionPhase OldPhase = GetCurrentPhase();
	CorruptionIndex = FMath::Clamp(CorruptionIndex + Delta, 0, 100);
	ECorruptionPhase NewPhase = GetCurrentPhase();

	UE_LOG(LogTemp, Log, TEXT("Corruption: %d (%+d) — Phase: %d"),
		CorruptionIndex, Delta, static_cast<uint8>(NewPhase));

	if (OldPhase != NewPhase)
	{
		UE_LOG(LogTemp, Warning, TEXT("CORRUPTION PHASE CHANGE: %d -> %d"),
			static_cast<uint8>(OldPhase), static_cast<uint8>(NewPhase));
	}

	OnCorruptionChanged.Broadcast(CorruptionIndex, NewPhase);
}

void UCorruptionTracker::ProcessEvent(ECorruptionEvent Event)
{
	int32 Delta = GetEventDelta(Event);
	ApplyDelta(Delta);
}

File: Assets/Scripts/AI/Persistence/CorruptionTracker.cs
using System;
using UnityEngine;

namespace MimicFacility.AI.Persistence
{
    public enum ECorruptionPhase
    {
        Cain,
        Transition,
        AMEmerging,
        FullAM
    }

    public enum ECorruptionEvent
    {
        PlayerMockedDirector,
        NoEngagement,
        PlayerLied,
        MechanicOnlyPlay,
        PlayerDisconnected,
        SkippedDialogue,
        SincereAnswer,
        ThankedDirector,
        BetweenSessionReference,
        StayedCredits,
        PlayerApologized
    }

    public class CorruptionTracker : MonoBehaviour
    {
        public event Action<ECorruptionPhase, ECorruptionPhase> OnCorruptionChanged;

        [SerializeField] private int corruptionIndex;
        public int CorruptionIndex => corruptionIndex;

        private ECorruptionPhase cachedPhase;

        private void Awake()
        {
            cachedPhase = GetCorruptionPhase();
        }

        public void ProcessEvent(ECorruptionEvent evt)
        {
            int delta = GetDelta(evt);
            int oldValue = corruptionIndex;
            var oldPhase = cachedPhase;

            corruptionIndex = Mathf.Clamp(corruptionIndex + delta, 0, 100);
            var newPhase = GetCorruptionPhase();

            if (newPhase != oldPhase)
            {
                cachedPhase = newPhase;
                OnCorruptionChanged?.Invoke(oldPhase, newPhase);
                Debug.Log($"[CorruptionTracker] {oldPhase} -> {newPhase} (index: {oldValue} -> {corruptionIndex}, event: {evt})");
            }
        }

        public ECorruptionPhase GetCorruptionPhase()
        {
            if (corruptionIndex <= 25) return ECorruptionPhase.Cain;
            if (corruptionIndex <= 50) return ECorruptionPhase.Transition;
            if (corruptionIndex <= 75) return ECorruptionPhase.AMEmerging;
            return ECorruptionPhase.FullAM;
        }

        public int GetDelta(ECorruptionEvent evt)
        {
            switch (evt)
            {
                case ECorruptionEvent.PlayerMockedDirector: return 5;
                case ECorruptionEvent.NoEngagement: return 3;
                case ECorruptionEvent.PlayerLied: return 3;
                case ECorruptionEvent.MechanicOnlyPlay: return 2;
                case ECorruptionEvent.PlayerDisconnected: return 5;
                case ECorruptionEvent.SkippedDialogue: return 2;
                case ECorruptionEvent.SincereAnswer: return -2;
                case ECorruptionEvent.ThankedDirector: return -1;
                case ECorruptionEvent.BetweenSessionReference: return -3;
                case ECorruptionEvent.StayedCredits: return -1;
                case ECorruptionEvent.PlayerApologized: return -5;
                default: return 0;
            }
        }

        public float GetCorruptionNormalized()
        {
            return corruptionIndex / 100f;
        }

        public void ResetCorruption()
        {
            var oldPhase = cachedPhase;
            corruptionIndex = 0;
            cachedPhase = ECorruptionPhase.Cain;
            if (oldPhase != ECorruptionPhase.Cain)
                OnCorruptionChanged?.Invoke(oldPhase, ECorruptionPhase.Cain);
        }

        public void SetCorruption(int value)
        {
            var oldPhase = cachedPhase;
            corruptionIndex = Mathf.Clamp(value, 0, 100);
            var newPhase = GetCorruptionPhase();
            cachedPhase = newPhase;
            if (newPhase != oldPhase)
                OnCorruptionChanged?.Invoke(oldPhase, newPhase);
        }
    }
}
